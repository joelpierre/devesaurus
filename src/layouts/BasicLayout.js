import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './Layout.module.scss';

import SEO from '../components/core/SEO/SEO';
import { getSiteMeta, getSiteOptions } from '../store/actions';
import PrimaryFooter from '../components/organisms/PrimaryFooter/PrimaryFooter';

export class BasicLayout extends PureComponent {
  componentDidMount() {
    const { onGetSiteMeta, onGetSiteOptions } = this.props;
    onGetSiteMeta();
    onGetSiteOptions();
  }

  render() {
    const { children, title, description, classes } = this.props;

    return (
      <>
        <SEO title={title} description={description} />
        <main
          data-test="basic-layout-main"
          className={classNames([styles['primary-main'], classes])}
        >
          {children}
        </main>
        <PrimaryFooter />
      </>
    );
  }
}

/* istanbul ignore next */
BasicLayout.defaultProps = {
  title: 'Default Title',
  description: 'Default description',
  classes: null,
  onGetSiteMeta: () => {},
  onGetSiteOptions: () => {},
};

BasicLayout.propTypes = {
  title: PropTypes.string,
  description: PropTypes.string,
  classes: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  onGetSiteMeta: PropTypes.func,
  onGetSiteOptions: PropTypes.func,
};

/* istanbul ignore next */
const mapStateToProps = ({ core: { title, description } }) => ({
  title,
  description,
});

export default connect(
  mapStateToProps,
  {
    onGetSiteMeta: getSiteMeta,
    onGetSiteOptions: getSiteOptions,
  }
)(BasicLayout);
