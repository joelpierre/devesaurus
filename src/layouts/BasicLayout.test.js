import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { BasicLayout } from './BasicLayout';
import {
  checkProps,
  findByTestAttr,
  matchSnapshot,
} from '../utils/test-utilities';

const defaultProps = {
  children: 'Hello',
  onGetSiteMeta: jest.fn(),
  onGetSiteOptions: jest.fn(),
};

/**
 * Factory function to create a ShallowWrapper for the BasicLayout component.
 * @param {object} props - Component props specific to setup
 * @returns {ShallowWrapper}
 */
const setup = (props = {}) => {
  const setupProps = { ...defaultProps, ...props };
  return shallow(<BasicLayout {...setupProps} />);
};

describe('<BasicLayout/>', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = setup();
  });

  it('renders the BasicLayout Component without errors', () => {
    const component = findByTestAttr(wrapper, 'basic-layout-main');
    expect(component.length).toBe(1);
  });

  it('should render with correct props', () => {
    checkProps(BasicLayout, defaultProps);
  });

  it('should match snapshot', () => {
    matchSnapshot(wrapper);
  });

  it('should get siteMeta and siteOptions on componentDidMount', () => {
    wrapper = setup({
      onGetSiteMeta: jest.fn(),
      onGetSiteOptions: jest.fn(),
    });

    const spy1 = wrapper.instance().props.onGetSiteMeta;
    const spy2 = wrapper.instance().props.onGetSiteOptions;

    wrapper.instance().componentDidMount();

    expect(spy1).toHaveBeenCalled();
    expect(spy2).toHaveBeenCalled();
  });
});
