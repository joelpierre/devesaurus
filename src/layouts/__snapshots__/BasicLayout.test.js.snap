// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BasicLayout/> should match snapshot 1`] = `
<Fragment>
  <SEO
    description="Default description"
    keywords={Array []}
    lang="en"
    meta={Array []}
    title="Default Title"
  />
  <main
    className="primary-main"
    data-test="basic-layout-main"
  >
    Hello
  </main>
</Fragment>
`;
